### ISSUE TITLE
Serious memory leak in R package

### ISSUE THREAD

**Issue by richardbeare** [SimpleITK/SimpleITK#131]
While images created by calls to the Image() constructors are garbage collected correctly, those returned from functions and filters are not:

```r
library(SimpleITK)

p <- replicate(5, Image(512, 512, 512, 'sitkFloat64'))
## check memory usage with ps
p <- 0
gc()
## check again - OK

## What about if we use a filter
InputIm <- Image(512, 512, 512, 'sitkFloat64')

ss <- ShiftScaleImageFilter()
ss$SetShift(1.0)
ss$SetScale(3)

pp <- replicate(5, ss$Execute(InputIm))
## check RAM
pp <- 0
gc()
## check again - memory leak

## Same problem with operations via functions?

ProcessedIms <- replicate(5, ShiftScale(InputIm, 1, 2))
## check RAM with PS
ProcessedIms <- 0
gc()
## check again - problem - doesn't change
```
Tracking this down now.

**Comment by blowekamp** [comment]
@zivy Have you encountered this problem? Is it new or has is existed for a while?

**Comment by zivy** [comment]
@blowekamp, @richardbeare 

I was able to reproduce the problem with R, OS X and linux (works correctly in Python). I am guessing this is a SWIG issue? The image destructor is not invoked when the image is the output of a filter. 

To see this I added a print statment to the image destructor:
std::cout<<"Deleting image"<<std::endl;

Add this at line 38 of sitkImage.cxx 

Below is the code I used:

```
library(SimpleITK)

img <-  Image(512, 512, 512, 'sitkFloat64')

# Using the object oriented filter interface
ss <- ShiftScaleImageFilter()
ss$SetShift(1.0)
ss$SetScale(3)

# the following line should show two images deleted (internal to the filter)
ss_image <- ss$Execute(img)

rm(img)
# the following line should show one deletion
gc()

rm(ss_image)
# the following line should show one deletion BUT DOESN'T
gc()

# Using the procedural filter interface
# the following line should show two deletions (internal filter images),
g_image <- GaussianSource()

rm(g_image)
# the following line should show one deletion BUT DOESN'T
gc()

```

I tried to do the memory monitoring in R as described in the [Advanced R book](http://adv-r.had.co.nz/memory.html), but this doesn't seem to work. All of the relevant functions (object_size, mem_used, mem_change) returned values that were too small. Looks like R is unable to determine the correct size of the memory allocations/deletions. I am not sure if this is related to the memory leak problem which is likely due to an internal filter variable still pointing to the allocated image?

**Comment by richardbeare** [comment]
The memory monitoring tools are all targeted at traditional R objects, not things allocated by external libraries. 

I can only assume that the problem has always been there. My error, I suppose. I was checking problems caused by disposing classes and assumed everything was OK once that worked as the swig-generated code looks the same.

I've sent a pull request that improves the situation, but is specific to images. I'd prefer a more general fix.

**Comment by blowekamp** [comment]
This has been demonstrated to be a problem with the sitk::Image type. I'd expect it to be a problem with the sitk::Transform type too. SimpleITK is fortunate in that we don't currently use a lot of different data types. But this looks like a super critical problem with the SWIG wrapping!

We also commonly use std::vector and std::lists etc... Do these types also have memory leaks?

**Comment by richardbeare** [comment]
std::vectors are converted directly to matching R data types, so I think this is OK, but will try to dream up a way of checking.

Transforms probably suffer in the same way as simple::Image.

I guess this is where the lack of a proxy layer causes a problem ...

**Comment by blowekamp** [comment]
The "fix" for this is targeted for a SimpleITK 1.0 patch release.

**Comment by blowekamp** [comment]
closed by #182
